### ISSUE TITLE
Please add LUT effects

### ISSUE THREAD

**Issue by AlexDaniel** [vanruesc/postprocessing#231]
The feature is perhaps best described by an existing ticket in three:
* https://github.com/mrdoob/three.js/issues/19457

There's a basic external [implementiation](https://github.com/gkjohnson/threejs-sandbox/tree/master/3d-lut) but it is perhaps not up to standards of this library :) For one, it doesn't use 3D textures which should at the very least result in less code (and be faster too, maybe?). That said, it does have a working loader for .cube files, so it's a great start.

Basically, LUTs provide a generalized way to perform any color operation or a combination of operations at the same time. They are extremely convenient when trying to create a specific look of a scene or the whole game.

**Comment by vanruesc** [comment]
I agree, a LUT-based color transformation effect would be nice!

 The reference implementation will come in handy, thanks for that. @gkjohnson Would it be ok if I used your LUT-implementation as a basis for a new `ColorTransformationEffect`?

I'd like to make this a WebGL 2 effect to avoid emulating 3D textures. The initial implementation will use linear color interpolation, tetrahedral interpolation will come later for reduced artifacts/banding: https://i.imgur.com/4SVNP.png

**Comment by gkjohnson** [comment]
> I'd like to make this a WebGL 2 effect to avoid emulating 3D textures.

If you're just going to use 3d textures then my implementation may not be all that useful because it was implemented with the intent to support WebGL1 -- a 3d texture LUT fragment shader should just be a couple lines. You're free to use it if it is helpful, though!

I'm planning to add that LUT implementation into three.js examples with a 3d texture sampling but it makes sense to have a version here, as well. I was going to conditionally support WebGL1 via 2d textures by testing the glsl version (or #define) and using different uniforms if you're interested in supporting webgl 1, too:

```glsl
#if __VERSION__ >= 300

    // 3d texture sample

#else

    // manual interpolation

#endif
```

I'll also add the 3DL and CUBE LUT loaders that output 3d textures into three.js so those should be useable by your implementation. I may be able to put some time into incorporating all that in the next couple weeks.

> The initial implementation will use linear color interpolation, tetrahedral interpolation will come later for reduced artifacts/banding: https://i.imgur.com/4SVNP.png

Regarding tetrahedral interpolation [this](http://www.ijetch.org/papers/318-T860.pdf) was the most clear description of how it works that I'd come across when I was looking into this -- unfortunately I couldn't find any reference implementations but that paper has a pretty nice description. In practice though I'm not sure how common using tetrahedral interpolation is for this, though? A larger LUT would be faster and probably "good enough" at removing noticeable artifacts if they're there -- a tetrahedral implementation would require more calculations and 4 texture samples vs just the 1 for the built in linear interpolation. If you have any other references related to this I'd be interested in checking them out.

**Comment by AlexDaniel** [comment]
> A larger LUT would be faster

But transferring that larger lut over the network can be a problem. For example, 33 point cube file is ≈0.7 MB (≈0.3 MB gzipped), but 65 point cube file is ≈5.8 MB (≈2.2 MB gzipped). It seems to me that some kind of interpolation has to happen, but not necessarily at every frame considering that LUTs are mostly static.

**Comment by vanruesc** [comment]
> You're free to use it if it is helpful, though!

@gkjohnson Cool, thanks! I haven't really delved into the matter yet, so it's still possible that I'll add WebGL 1 support. It just feels wrong to keep supporting and promoting legacy tech. Heck, even WebGL 2 feels outdated at this point with WebGPU on the horizon.

@AlexDaniel right, there are a lot of people who have to deal with low bandwidth, myself included, not to mention mobile users. Tetrahedral interpolation would add a nice option to trade performance for fidelity.

> If you have any other references related to this I'd be interested in checking them out.

@gkjohnson I've only found [this shader implementation](https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/master/src/OpenColorIO/ops/lut3d/Lut3DOpGPU.cpp) so far. Haven't taken a closer look at it yet, though.

> It seems to me that some kind of interpolation has to happen, but not necessarily at every frame considering that LUTs are mostly static.

@AlexDaniel Good point. However, I think the shader would need to interpolate the LUT anyway, so the question is whether a LUT augmentation on the CPU would be worth it compared to just doing tetrahedral interpolation in the shader. OpenColorIO also has a [CPU implementation](https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/master/src/OpenColorIO/ops/lut3d/Lut3DOp.cpp).

**Comment by AlexDaniel** [comment]
> However, I think the shader would need to interpolate the LUT anyway, so the question is whether a LUT augmentation on the CPU would be worth it compared to just doing tetrahedral interpolation in the shader.

Well, the two are not necessarily mutually exclusive, maybe both shaders can be useful (and be used at the same time, actually). That is, maybe tetrahedral shader can be used to create a much larger 3D texture, and then the fast shader can be used at every frame. But maybe that's making things a bit more complicated. Anyway, please judge yourself.

**Comment by vanruesc** [comment]
After reading [these notes](https://docs.unrealengine.com/en-US/Engine/Rendering/PostProcessEffects/UsingLUTs/index.html#additionalnotes), it looks like LUTs are not the best way of doing color grading. Unreal is phasing this feature out in favor of individual color grading properties for things like white balance, shadows, midtones and highlights.

I think it might be better to implement a proper tonemapper and add more color grading options similar to what [Unreal is describing here](https://docs.unrealengine.com/en-US/Engine/Rendering/PostProcessEffects/ColorGrading/index.html#workflowforgoodcolorcorrection). I wanted to move auto-exposure and tonemapping into separate effects anyway.

**Comment by AlexDaniel** [comment]
> After reading these notes, it looks like LUTs are not the best way of doing color grading. Unreal is phasing this feature out in favor of individual color grading properties for things like white balance, shadows, midtones and highlights.

Actually, LUTs end up being the only way to do color grading. Of course, it is always possible to create some very primitive adjusting sliders and call that color grading tools, but most colorists will find that to be insufficient if not outright useless.

I don't know much about the internals of unreal, but I wouldn't be surprised if they are generating a single LUT based on the user input in their color grading tools, and then just using that in the shader. If they're not doing that yet, then eventually they will once they make their tools more sophisticated (with things like custom rgb, hue, saturation and other *curves*, color selectors, etc.). Basically, because they provide their own GUI, they can afford to reinvent color grading software so that they can bypass the LUT format, which allows them to handle HDR a bit better. That seems to be the point being made, but it is a different realm which doesn't exactly apply to three.js. Maybe in case of babylon.js the approach would've been different.

In other words, LUT implementation is still needed. Whether the LUT will be created automatically (taking into account the availability of HDR and other things) or externally by the user is a different question. I guess for now the best idea is to focus on artistic freedom and to just support LUTs. This probably means that those who want to generate HDR-capable content will need to create at least 2 different LUTs, which isn't a bad idea given that HDR to SDR conversion is not a well-defined process and may involve some creative decisions too. But yes, it is kinda unfortunate that people would need to put extra work to get HDR results.

**Comment by gkjohnson** [comment]
Bandwidth consideration are a good point and "upscaling" a smaller LUT using tetrahedral interpolation seems like a reasonable solution. Regarding performing this on the GPU as far as I'm aware you can't render directly into a 3d texture in WebGL -- in the future compute shaders may be a good candidate for this. This PR (https://github.com/mrdoob/three.js/pull/20111) adds support for rendering into layers of a 3d texture but it's not yet supported by three.js. Performing the upscaling in javascript might not be so bad performance wise but I'd have to try it. Sounds like something that can come later though.

> I haven't really delved into the matter yet, so it's still possible that I'll add WebGL 1 support. It just feels wrong to keep supporting and promoting legacy tech. Heck, even WebGL 2 feels outdated at this point with WebGPU on the horizon.

I definitely feel this. Unfortunately Safari has been dragging its feet in terms of adding WebGL2 support so dropping WebGL1 means dropping support for all Apple devices until it's there. They've apparently committed to adding it now but it's unclear what the timeline is. Hopefully WebGPU isn't so slow to be adopted fully but who knows.

> I think it might be better to implement a proper tonemapper and add more color grading options similar to what Unreal is describing here.

Interesting -- it sounds like it's motivated primarily to support HDR monitors, though, which I don't know if is something the web even really is talking about supporting. Either way it seems they're still a useful tool particularly for LDR content for controlling a feeling of the final image that can be produced by a variety of existing software.

**Comment by gkjohnson** [comment]
> Not exactly. From what I understand, Safari allows you to enable WebGL 2 in experimental settings.

It does now but I believe that flag has been around for years and Firefox and Chrome have had full support [since early 2017](https://caniuse.com/?search=webgl2) -- Safari has been extremely slow to add WebGL2 and while they've apparently committed a bit more firmly now and I'd like to believe they will have a support in a few months I'm not really betting on that. A nag screen may be good enough for your app but when building something for general use the uncertainty about WebGL2 support is something to consider.

**Comment by vanruesc** [comment]
Thanks for the feedback.

I see what you mean; LUTs are indeed better in terms of practicality - especially since we don't have a visual game editor where we could easily tweak individual color grading settings. However, I think it makes sense to also add more atomic color grading effects that operate on linear colors because those are more portable from a technical perspective.

From what I've gathered, LUTs operate in the sRGB color space and need to be applied at the very end of the rendering pipeline. Furthermore, LUT color grading cannot be implemented as an `Effect` because effects run before the color encoding step. LUT support should instead be implemented as a feature of the `EffectMaterial`; If a LUT is provided, it will be applied [after the color encoding step](https://github.com/vanruesc/postprocessing/blob/4fbd2bdb4179cbf142154af16d28e6c76cd8da7b/src/materials/glsl/effect/shader.frag#L73). And the `EffectComposer` should probably complain if a LUT is used by an `EffectPass` that doesn't render to screen.

That's the plan so far, did I get anything wrong?

**Comment by AlexDaniel** [comment]
> From what I've gathered, LUTs operate in the sRGB color space and need to be applied at the very end of the rendering pipeline.

That doesn't sound quite right. Textures and other things are also often in sRGB, and it just means they need to get converted (`.encoding = sRGBEncoding`) but otherwise everything can still work with linear colors.

So, my expectation is that I'd be able to apply a LUT at any point in the EffectComposer, and even multiple times (with different LUTs) if so is required. I agree that this means that a regular sRGB LUT then needs to be converted to linear colors. For values of the LUT, `.encoding = sRGBEncoding` should be enough, I think. For the lookup itself a bit of extra glsl is required to pick the right values in the LUT given a linear color. Another option is to convert it ahead of time. Does it make sense, or is there any reason why an sRGB LUT can't be converted to linear colors or used with linear colors in realtime?

Please keep in mind that this whole GLSL business is a new territory for me so I'm kinda talking out of my ass :)

**Comment by vanruesc** [comment]
> is there any reason why an sRGB LUT can't be converted to linear colors or used with linear colors in realtime?

Afaik, LUTs map a finite range of LDR sRGB input colors to LDR sRGB output colors. Linear color values don't have a limited range which makes them incompatible with LUTs. The Unreal docs explain that a LUT must be applied after tonemapping which reduces colors to [0.0, 1.0]. Since the sRGB conversion is the last step in the pipeline, I think it would make sense to apply the LUT afterwards.

I don't know if converting a LUT into linear space would work out. Applying a converted LUT transformation to linear colors and then converting the result into sRGB might produce incorrect results.

I believe the point of a LUT is efficiency, so using multiple LUTs seems a bit counterintuitive.

**Comment by vanruesc** [comment]
I think I'm just confused about HDR/LDR and tonemapping in general. Will need to study this more.

**Comment by AlexDaniel** [comment]
@vanruesc I also don't understand it fully, if at all. Here's what I understand. If we take ACES tone mapping for example, then in color grading terms it's just an S-curve. Essentially, that's a LUT too. Almost any grade of any video source includes an S-curve at some point, otherwise the video will look too “flat” (unless the camera itself overdoes it for you, which doesn't really happen with prosumer and up hardware).

Now, in gamedev world, the output from 0 to 16 (or whatever it is that corresponds to the max outputable value) is scaled down into 0 to 1 range, which makes things look flat, so an S-curve is slapped on top of that, which makes things look alright or “filmic”. Boom, ACES.

However, unless I'm using some kind of adaptive tonemapping, is there even any reason to have an ACES curve step if I can include an S-curve as part of my grade? That is, why not include an S-curve in the LUT itself.

Or, in other words, why can't tonemapping itself be an Effect? Yes, in most cases you'd want to place it last, but hey, who knows.

So, yeah, I don't understand it. Please let me know where exactly I got confused.

> I believe the point of a LUT is efficiency, so using multiple LUTs seems a bit counterintuitive.

While it's true that you can merge multiple consecutive LUTs into one, that doesn't really work if you want to have any temporal or spatial effects in between your LUTs (let's say before bloom and after bloom tweaks). Moreover, having a single base grade for the whole project and additional ones for specific scenes can be very useful too. So, you're right, but being able to use one LUT only can be very limiting.

**Comment by AlexDaniel** [comment]
From [unity docs](https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@6.9/manual/Post-Processing-Tonemapping.html):

> Tonemapping Mode:
>  • Neutral: Use this option if you only want range-remapping with minimal impact on color hue & saturation. **It is generally a great starting point for extensive color grading.**
> • ACES: Use this option to apply a close approximation of the reference ACES tonemapper for a more filmic look. It is more contrasted than Neutral and has an effect on actual color hue & saturation. Note that if you use this tonemapper all the grading operations will be done in the ACES color spaces for optimal precision and results.

So, I'm not wrong. One can use range remapping + LUTs for “extensive color grading”. That sounds about right.

**Comment by gkjohnson** [comment]
@vanruesc I've added PR into three.js that adds LUT support for postprocessing here:

https://github.com/mrdoob/three.js/pull/20558

It supports both 3d texture and 2d texture luts pretty seemlessly. And I've updated the 2d texture LUT sampling so that it exactly matches the 3d texture linear interpolation sampling. Feel free to grab what you want from it!

Regarding HDR support -- Unity's docs might be worth looking at:

https://docs.unity3d.com/Packages/com.unity.postprocessing@2.1/manual/Color-Grading.html

In order to support color grading HDR renders they require a custom log-encoded LUT but I'm not how easy those are to produce. It seems the common case is to just apply the LU on the LDR, post-tonemapped result.

**Comment by vanruesc** [comment]
Got it, thank you!

**Comment by AlexDaniel** [comment]
> In order to support color grading HDR renders they require a custom log-encoded LUT but I'm not sure how easy those are to produce. It seems the common case is to just apply the LUT on the LDR, post-tonemapped result.

In the page you linked, they talk about their color grading effect which creates the LUT internally. They also provide “Lookup Texture” which is applied “before the rest of the color grading operators”.

> It seems the common case is to just apply the LUT on the LDR, post-tonemapped result

Doesn't seem to be so simple to me, especially given that the tonemapping itself that they do appears to be part of the generated LUT.

**Comment by vanruesc** [comment]
Sorry for the delay; I haven't been able to work on this yet. Hopefully this weekend I'll find some time.

**Comment by vanruesc** [comment]
Made some progress, but it's not done yet. I'll give another update later this week.

**Comment by vanruesc** [comment]
I've been too busy last week to finish this, but it shouldn't take much longer. I'll try to get it done before the end of the month.

**Comment by vanruesc** [comment]
Sorry for the repeated delays - work takes priority at the moment.

The `dev` branch now contains the loaders which return a `LookupTexture3D` instead of a normal 3D data texture. This allows to convert the texture into a 2D strip LUT. Converting from sRGB to linear is also supported, but I still need to test everything thoroughly.

I ended up naming the effect `ColorGradingEffect` because it will eventually support baking arbitrary color grading operators into a single composite LUT. This would also replace other existing color grading effects in the library.

Tonemapping could be included in the LUT but I think this would complicate things too much right now as it requires a special input domain other than `[0, 1]` and probably some form of encoding. So for now, I think it's best to treat tonemapping separately.

**Comment by vanruesc** [comment]
Added `LUTEffect` and related classes in `postprocessing@6.19.0`.

I changed the name of the effect again because it would make more sense to apply color grading operations directly to the LUTs if those are ever added.

#### Some closing notes

This turned out to be a can of worms in regards to color space conversions and LUT formats and took a lot longer than anticipated. There's always a chance that some bugs made it into the first release so let me know if you run into any issues.

The final implementation relies on `LookupTexture3D` to perform various LUT transformations (sRGB<->linear, Uin8<->float, 3D<->2D). It also supports converting plain image LUTs into 3D textures since those are more commonly found in the wild.

I modified the parsing methods of the loaders to use regular expressions for the iteration which made things a little bit faster:

| | LUT3dlLoader | LUTCubeLoader |
|-|-|-|
| __original__ | 21-31ms | 20-27ms |
| __new__ | 15-23ms | 14-19ms |

I also moved the tetrahedral upscaler into a web worker and removed all references to three to avoid including the library in the worker. For that, I converted the object-based vector math into atomic operations which resulted in a major speedup:

| | TetrahedralUpscaler (64) | TetrahedralUpscaler (128) |
|-|-|-|
| __original__ | 295-300ms | 2243-2272ms |
| __new__ | 115-134ms | 817-886ms |

@gkjohnson All of this would've taken a whole lot longer if it weren't for your reference implementation, so thank you! If you'd like to use any of my modifications feel free to do so. No attribution needed.

**Comment by AlexDaniel** [comment]
@vanruesc this is fantastic! Thank you so much! My project got stalled a little bit due to some unexpected things, but I'm slowly getting back to it now. Will provide some feedback as soon as I have it!

**Comment by AlexDaniel** [comment]
As a quick feedback, the `scale up` option in the demo (https://vanruesc.github.io/postprocessing/public/demo/#color-grading) seems to be a bit off? Or maybe everything is a little bit off. Let's see:

For example, let's choose “neutral” lut (I'm assuming it's a bypass lut that shouldn't affect colors at all, is that right?).
* If “scale up” is not selected, then changing the opacity from 0 to 1 changes the resulting image. That is, the bypass lut is not bypassing the colors correctly. Yes, the difference is not huge, but there is a difference that shouldn't be there. Moreover, the image overall seems to become darker, so maybe the inaccuracy goes one way only.
* If we now choose `scale up` with the highest available size (128), then the colors change once again. OK, in case of extreme luts that can indeed happen, but we're not using an extreme lut in this case, the colors should be 100% the same. When looking more closely, I can see some color banding:
![image](https://user-images.githubusercontent.com/5507503/103483061-572bac80-4ded-11eb-960f-30537328780d.png)
Compared with opacity 0:
![image](https://user-images.githubusercontent.com/5507503/103483077-70ccf400-4ded-11eb-8d4a-5f4942e63ae4.png)
* On the other hand, smaller luts (target size 32) seem to have no banding. Hmm?

So, I don't know where exactly it goes wrong, but it seems like there are some tiny issues somewhere.

**Comment by vanruesc** [comment]
@AlexDaniel I've seen these artifacts too during development but I think it's simply a shortcoming of LUTs and interpolation in general. I have double checked the TetrahedralUpscaler as well as the LUT shader and found no bugs there. I have also downloaded @gkjohnson's original implementation and plugged in a screenshot of the Sponza demo scene to cross-check for differences but it produced the same results. If you focus on the darker areas in the [original demo](https://gkjohnson.github.io/threejs-sandbox/3d-lut/), you can see that the neutral LUTs also makes the image darker.

I've read somewhere that LUTs heavily depend on the input colors  and that results will be bad if the source image is not properly color corrected. I'm no expert in this field but it seems that the Sponza demo is simply too dark and thus suffers mild color degradation when put through the LUT filter.

I'm taking a wild guess here but the increase in banding artifacts when scaling up the LUT might be linked to a higher error rate of smaller linear interpolations performed in the shader. Recall this comparison: https://i.imgur.com/4SVNP.png There's a good chance that tetrahedral interpolation could solve the banding issue if performed inside the shader. But it would obviously require more GPU cycles.

> On the other hand, smaller luts (target size 32) seem to have no banding. Hmm?

The base size of the cinematic 3dl/cube LUTs and the neutral LUT is 32 so the upscaling doesn't change them if the target size is 32.

**Comment by AlexDaniel** [comment]
> I've read somewhere that LUTs heavily depend on the input colors and that results will be bad if the source image is not properly color corrected.

This isn't a technical limitation. LUTs create a certain look&feel for the image you're seeing, and that of course works better if the input is normalized (a simple example: push darker colors into teal to create a sort of filmic look – if your image is too dark then the whole image will be teal, and that's not what you wanted). So usually one would do color correction first and color grading later. The difference is in the semantics and human expectations, both steps are done with LUTs.

> I think it's simply a shortcoming of LUTs and interpolation in general

Download the free version of Davinci Resolve, add any source image, go to color grading and load the neutral cube lut. Apply it to the image. I can pretty much bet that there will be no change in color whatsoever. They do it with shaders, too.

**Comment by gkjohnson** [comment]
> If “scale up” is not selected, then changing the opacity from 0 to 1 changes the resulting image. That is, the bypass lut is not bypassing the colors correctly. Yes, the difference is not huge, but there is a difference that shouldn't be there. Moreover, the image overall seems to become darker, so maybe the inaccuracy goes one way only.

If "scale up" is not selected then basic linear interpolation is used for sampling so you'll have to expect some color differences. If you want better color accuracy you'll have to upscale the LUT. I think there are still improvements to be made, though.

@vanruesc 

I did find one issue in my TetrahedralUpscaler class which was causing some banding on upscale -- specifically the UVW values were not being adjusted to be relative to the sub 2x2x2 pixel cube that you're actually sampling from. The commit that fixes it is [here](https://github.com/gkjohnson/threejs-sandbox/commit/a4f35f1ec618c2297a2027390ae95a7e0fb7bed9). For testing I adjusted my demo to use a `FloatType` format for the render target to ensure no color banding would occur due to gamma correction and it looks much better with the fix. Upscaling the neutral 2³ LUT in my example to 256³ looks just about identical (if not identical) to the original image though there are more noticeable differences when upscaling 32³ to 256³.

It occurs to me that there may be other issues with sampling in the shader and in the upscale that could be addressed but I want to think about them more -- perhaps you'll have some thoughts. Right now the 3d texture LUT is set to edge clamp for sampling and the color is used directly as the UVW coordinate for sampling which means that every color below half a pixel width will be clamped to the edge color. It seems likely that the right thing to do is to start sampling [0,1] from the center of the edge pixels rather than the edge of the texture ie inset the sampling by half a pixel. I think this should be accounted for in both the shader and in the upscaler. Let me know if I'm not being clear. When I have time I'll mull it over a bit more, too.

**Comment by AlexDaniel** [comment]
Maybe I am missing something, but in my view it is not useful to hold the position that the LUT is allowed to produce completely wrong results. Yes, maybe a few pixels on the whole image can be off by 1, but that shouldn't be noticeable to the human eye at all. I don't know what will convince you, but here is an example of somebody applying a small neutral lut with the size of 16 and producing results that are indistinguishable from the original: https://kosmonautblog.wordpress.com/2017/04/26/color-grading-correction/. Testing is very simple – if you can see *any* difference after applying the lut, something is wrong.

**Comment by AlexDaniel** [comment]
By the way:

> The base size of the cinematic 3dl/cube LUTs and the neutral LUT is 32 so the upscaling doesn't change them if the target size is 32.

Looking at the repo, I found this: https://github.com/vanruesc/postprocessing/blob/main/public/demo/textures/lut/lut-neutral.jpg

But that lut seems to be of size 16, or am I wrong? Maybe there is actually no problem and using an actual lut with size 32 will give proper results.

**Comment by vanruesc** [comment]
> I don't know what will convince you

I didn't mean to downplay this issue - just trying to find an explanation. Let's track this bug over here #250 and fix it.

That neutral LUT is not actually used in the demo, I just forgot to delete it.

@gkjohnson thanks for the heads-up! I'll update the TetrahedralUpscaler when I get a chance.

**Comment by AlexDaniel** [comment]
I also didn't mean to sound too frustrated, I'm loving this library so far. :) It's just that *slight* unexpected color differences can be very annoying. For example, see [this](https://blog.maxofs2d.net/post/148346073513/bt601-vs-bt709).
